File: code/78-Subsets/78-Subsets.py
# -*- coding: utf-8 -*
"""
LeetCode 78 - Subsets

解题步骤（自动整理）
1. 用 DFS/回溯枚举所有可能选择，并维护当前路径/状态
2. 在递归过程中进行剪枝（如边界、重复、提前失败条件）
3. 到达终止条件时收集结果或返回，并回溯撤销选择
"""
r"""
@author: Kaka Chen
@license: (C) Copyright Personal Uncommercial Limited.
@contact: kakachen1990@gmail.com
@file: 78-Subsets.py
@time: 2020/9/6 1:06 下午
@desc:
Given a set of distinct integers, nums, return all possible subsets (the power set).

Note: The solution set must not contain duplicate subsets.

Example:

Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]

"""


class Solution(object):
    # def subsets(self, nums):
    #     """
    #     :type nums: List[int]
    #     :rtype: List[List[int]]
    #     递归
    #     """
    #     n = len(nums)
    #     result = [[]]
    #     for num in nums:
    #         result += [curr + [num] for curr in result]
    #     return result

    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        回溯
        """
        result = []
        n = len(nums)
        for k in range(n + 1):
            def backtrack(first=0, curr=[]):
                if len(curr) == k:
                    result.append(curr[:])
                for i in range(first, n):
                    curr.append(nums[i])
                    backtrack(1 + i, curr)
                    curr.pop()
            backtrack()
        return result


if __name__ == '__main__':
    s = Solution()
    ipt = [1, 2, 3, 4]
    print(s.subsets(ipt))

File: code/78-Subsets/78-Subsets.cpp
/*
LeetCode 78 - Subsets

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
/*Subsets
Given a set of distinct integers, S, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If S = [1,2,3], a solution is:

[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]

妈的网上和教材里说的每一个对的，数位法要纸上画一下后再做，就会比较明白到底怎么回事
*/

class Solution {

public:
    vector<vector<int> > subsets(vector<int> &S) {
          sort(S.begin(), S.end()); //输出有序
          vector<vector<int> > res;
          long n = pow(2,S.size());//总共有n种情况

          for(long i=0;i<n;i++){
              vector<int> v;
              for(int j=0;j<S.size();j++){
                  if((1<<j)&i){
                      v.push_back(S[j]);
                  }
              }

              res.push_back(v);
              v.clear();
          }

          return res;

    }
};
